Competency Mastery Concurrency ADR (#657)
openedx/openedx-learning: docs/openedx_learning/decisions/0005-competency-mastery-storage.rst

.. _openedx-learning-adr-0005:

5. How should learner competency status be stored at scale?
===========================================================

Status
------
Proposed.

Context
-------
Per :ref:`openedx-learning-adr-0002`, competency achievement criteria form a boolean tree. An
internal ``CompetencyCriteriaGroup`` node combines child nodes with an ``AND``/``OR``
``logic_operator``, can be scoped to a course run, and can nest under a parent group. A
``CompetencyCriterion`` leaf is the tree's terminal node: it points at one tag/object association
and a rule.

The student mastery statuses tied to these tree nodes are stored in:
- `StudentCompetencyCriteriaStatus` (leaf nodes)
- `StudentCompetencyCriteriaGroupStatus` (middle nodes)
- `StudentCompetencyStatus` (top-level)

For each of these, we also need to persist history, because we need an audit trail to understand
why a learner did or didn't achieve mastery of a particular competency or any of the associated "measurement instruments"

> (gradeable subsections).

Reviewer: It does not block this ADR, but let's talk with Product/UX folks and explore just how much of an audit history we really need. History storage is the most dangerously unbounded thing in this system. Moving it to a separate system and potentially pushing it to another database entirely doesn't really solve it.

For instance, say we determine that we only need history showing when they cross the aggregate threshold for the first time. We could potentially save billions of rows of storage for a large site that would otherwise be storing a new row for every time someone clicked "submit" on a problem somewhere.

Author: @mgwozdz-unicon what are your thoughts?

Author: Yes, I think that is already the intention here, let me know what you think:
status advances are only monotonic, e.g. from partially demonstrated to demonstrated. So whenever there is no such status advance, we will not write anything into the history table.
I even gave it a hard db constraint in ADR 0002:
"   7. ``StudentCompetencyCriteriaStatusHistory(user_id, competency_criteria_id, status_id)`` (unique -- at most one HISTORY row per learner, leaf, and status level, which also serves as the idempotency key for the append in :ref:`openedx-learning-adr-0004`)"
I will make this a separate point so that it is more clear.
Think like this: The active table stores group and leaf nodes - for each learner for each subsection they attempt, for example. Sure, that is a lot.
Now when the learner tries a subsection multiple times and never crosses the grade threshold, it will remain "attempted, not demonstrated" and only one history entry will be written - because there is only a one-time status advance.
Since every node just has 3 or 4 statuses - for example "not demonstrated" -> "partially demonstrated" -> "demonstrated" for group nodes (I'm paraphrasing here, it might be a little inaccurate), the history table will only ever receive 3 or 4 writes for the group node.
So the history table will only ever be 3-4x as large as the active table.

What are your thoughts so far? Does that alleviate your concern, or do you see further opportunity to cut down the history table?

Also, one note: if we want to allow course authors or django admins to manually edit the status later downwards in case there was a mistake, and we want to keep that in the audit trail as well, we need to drop the unique constraint.

Reviewer 2: I'm pretty sure some people will definitely want to see the history at that granular of a level, and in a few different use cases.

One of the things I remember from our interviews with institutions was that many folks were talking about the importance of learners being able to see their historical progress of competency attainment across multiple attempts in a single assignment as well as across the entire course, and surely at a program level as well. 

But there will also be analytics/reporting use cases where instructors and instructional designers will want to understand some kind of aggregation of how learners were progressing through each material as it rolls up to the competency so that they know which learning materials need improvements and can try to get insights about which learning materials did well so that they can emulate that success.

But yes, the product folks and our CBE SME Jason could probably all provide more insights here.  I'm sure that there are certain metrics that would take priority over others, but I feel pretty confident that auditing historical records will be desired at the leaf/subsection level, which is the one that is the biggest storage challenge. It's actually some of the middle roll-ups that some people may not care about as much, but I also think that some people will definitely want all the detail they can get. 

If someone has a reasonably complicated logical hierarchy for how to demonstrate mastery of a particular competency and a learner inquires about why they haven't received their badge for it, then someone will need to do an audit to figure out where in that logical hierarchy the student didn't achieve so that they can explain this to them. 

To Dave's point later on, there probably is a limit to the amount of time that a learner would ask the question "I thought I demonstrated mastery of this competency. Do an audit to find out why I was told I haven't." Which could probably make it safe to truncate the table at some point in history, though I know there are also laws and retention periods and such that need to be adhered to.

Out of these use cases, the one that would likely want historical data furthest in the past is actually probably the analytics/reporting case in the event of a course that runs over several months and they want to compare data for a specific subsection from the last 5-10 times this course took place, for example.

Storing every leaf multiplies out at scale. A course can carry on the order of 200 leaf criteria,
so the leaf level is where the row count concentrates: the leaf table (learners x attempted
leaves) potentially reaches the low billions for an Open edX instance with millions of learners. The dominant
multiplier is this per-leaf breadth (roughly 200x per course), not time. Mastery is monotonic (see
"Advance-only banking" below): a node can only advance through the small status lattice, at most a handful of
forward steps ever.

That scale is not, on its own, what makes a relational database struggle. A point lookup against a
billion-row table backed by the right composite index is a logarithmic-time index seek regardless

> of the table's size; the dashboard reads this feature performs are exactly such point lookups. What

Reviewer: ```suggestion
of the table's size; the dashboard that reads this performs exactly such point lookups. What
```

Is this what was meant?

billions of rows makes painful is schema migrations, backups, and any non-indexed or aggregate
query.

Decision
--------

**Alleviate performance concerns by following established edx-platform practices.**
edx-platform has already large tables of similar magnitude, such as `StudentModule`,
and they have proven themselves.
This means that as long as we follow their structure, we can have confidence that we are
not introducing any huge new problems.
They do not use physical partitioning, but work by using 64-bit primary keys,
enabling optional read-replica offloads, and splitting out a history table that
uses a dedicated database router. The decisions below are designed to mirror these

> existing patterns.

Reviewer: Okay, the database router thing was a giant, ugly hack, because I threw together the original `StudentModuleHistory` model in a half-day hackathon project to help out one of the first CS courses on the platform, and I neglected to override the primary key defaults to make it 64-bit. The result was that we were running out of primary keys, and some devops folks (Kevin I think?) retrofitted a new table in there

Also, the `courseware_studentmodulehistory` table stays somewhat manageable for large instances because they truncate it from time to time, because this history is used for course staff debugging purposes, not auditing. So this week's history is really important when a student is asking, "I submitted the right answer, why is the grader marking me wrong?", but that's not a thing we need to look up six months from now.

rows. Each level also has a parallel append-only HISTORY table that records one row per genuine
status advance, for audit and point-in-time reconstruction. Because status only advances, the status
at any past time is the latest recorded advance at or before that time, so point-in-time is fully
reconstructable from HISTORY. Because status is monotonic, the number of advances per node is bounded
by the status lattice (a small constant), so HISTORY grows with learners and nodes, not with time,
and stays about the same order of size as ACTIVE. Keeping ACTIVE and HISTORY separate still pays:
ACTIVE is a single in-place current row optimized for the dashboard point lookup and is the row
per-learner concurrency is anchored on (:ref:`openedx-learning-adr-0004`), while HISTORY is
append-only.

**64-bit primary keys from the start.** The leaf ACTIVE and HISTORY tables use a 64-bit
``BigAutoField`` primary key, chosen up front, mirroring edx-platform's
``UnsignedBigIntAutoField`` on ``PersistentSubsectionGrade`` ("primary key will need to be
large for this table"). Changing a primary-key type on a billion-row table later is

> prohibitively expensive.

Reviewer: This shouldn't need to be justified, as `BigAutoField` should be the default practice per [OEP-38](https://docs.openedx.org/projects/openedx-proposals/en/latest/best-practices/oep-0038-Data-Modeling.html#use-a-numeric-primary-key).

Also, please don't put up anything to imply that `UnsignedBigIntAutoField` is a good idea. It was an overkill reaction to having 32-bit primary keys run out on key tables, but it's more headache than it's worth. It theoretically gives us 2X the space of `BigAutoField`, but (a) the existing space of `BigAutoField` is so big that we'll never realistically reach it; (b) reaching it would break the database in a bunch of other ways anyway; and (c) declaring a custom field type like that introduces maintenance headaches like... (d) painful compatibility issues with PostgreSQL, which does not support unsigned ints.

**A dedicated database alias and router for the leaf HISTORY table, baked in from the start.**
This only applies to the `StudentCompetencyCriteriaStatusHistory` table.
The leaf HISTORY table (``StudentCompetencyCriteriaStatusHistory``), the dominant table in this
model, is the one learner-status table assigned to a dedicated Django database alias through a
database router, mirroring edx-platform's courseware-history router
(``StudentModuleHistoryExtended``), which is likewise a history table. The alias defaults to the

> main database.

Reviewer: Please don't do this. See my comment above about why this routing exists. Let's keep it simple and in the same database so that transactions just work.

**No database-level foreign keys to `user_id` on ACTIVE or HISTORY table.**

> Foreign keys to ``user_id`` must have ``db_constraint=False`` set, mirroring edx-platform's own

Reviewer: That was done because we were forced to in order to do an in-place upgrade courseware_studentmodule to have 64-bit primary keys without downtime on edx.org: openedx/openedx-platform#20971

It is not intended as a general pattern. Please leave a real foreign key.

``StudentModule``. This follows the app-boundary decision in :ref:`openedx-learning-adr-0001`, which keeps

> learner-status models decoupled from the concrete user model. Furthermore,

Reviewer: I'm not clear as to what part of the referenced ADR this is drawing from, but our convention in this repo has been to have real foreign keys to the User model—it's just that we respect Django's ability for projects to define custom user models by grabbing it from `settings.AUTH_USER_MODEL`, instead of assuming that it's always `django.contrib.auth.models.User`.

Author: Maybe my misunderstanding. https://github.com/openedx/openedx-core/blob/845dc1fc64f93023d7c5672ddc3e03c44effc2aa/docs/openedx_learning/decisions/0001-competency-criteria-location.rst?plain=1#L48

a real constraint has an ongoing cost at this table's write volume:
a hot user row would see extra lock contention under concurrent writes. This
follows the app-boundary decision in :ref:`openedx-learning-adr-0001`, which keeps

> learner-status models decoupled from the concrete user model.

Reviewer: So in theory, very little should write to the User model, but it's true that we have seen reports of concurrency issues that we don't understand very well at this point:

* https://github.com/openedx/completion/pull/443
* https://github.com/openedx/openedx-platform/pull/38815

On that note, if there is competency information that makes sense to attach in a 1:1 way per student, e.g. a `CompetencyStudent` model that would go 1:1 with `User`, it might make sense to make various competency models foreign key to that instead of directly to `User`.

In either case though, I think we can remove this decision from the ADR.

This is independent of the delete-protection boundary in :ref:`openedx-learning-adr-0002`
(Decision 7): that boundary keys off ``competency_criteria_id`` and its ancestor tables, not
``user_id``, so dropping the database-level constraint here does not weaken it.

**Enable read-replica offload for heavy reads for the leaf tables.**
This only applies to the ACTIVE `StudentCompetencyCriteriaStatus` and HISTORY
`StudentCompetencyCriteriaStatusHistory` tables. Offload is scoped to the leaf level because that is
where row count and read volume concentrate (the ~200x per-course leaf multiplier above); the group
and top-level status tables are orders of magnitude smaller, so their dashboard reads are cheap point
lookups that do not need to leave the primary and are not worth the replica-lag complexity.

> Prior art: ``edx_django_utils``'s ``read_replica_or_default()``.

Reviewer: This is always a possibility for the future, but one we'd want to punt until we had verified that it was a problem worth tackling and testing for. StudentModule load is really high because it's used in many, many places, the payload can be enormous, and the lookup pattern happens via a very expensive/large composite index. My hope would be that even on sites that make heavy use of CBE, the load from it should be much lower.

**Advance-only banking, monotonic.** Once a node reaches ``Demonstrated`` its ACTIVE row is retained
("banked"): the recorder never automatically regresses it, not on a later downward grade correction

> and not on a criteria change. This applies at every level, including the leaf. A genuine downward

Reviewer: I strongly suspect that this will end up changing in the future, but I'm okay with it as a starting position.

grade correction does not advance the status, so it writes no HISTORY row and leaves the banked
ACTIVE status unchanged; because HISTORY records only advances, it never carries suppressed
regressions. Reversing a banked status is a separate administrative action, out of scope here.
This monotonicity is what makes out-of-order and duplicate delivery safe, since a late or replayed
event can never lower a status, and :ref:`openedx-learning-adr-0004` relies on it.

**Retroactive criteria changes are monotonic for the learner.** A retroactive edit can newly grant
or preserve mastery, but it never silently revokes it, and it never rewrites a learner's recorded

> leaf mastery downward.

Reviewer: Again, I take the point that we shouldn't silently do this by default, but I expect people will have errors in their content settings, and this will need to be adjustable at some point. But also again, I'm not going to block this as a starting position.

Also, we've had mention of use cases in the past where mastery has a shelf-life, e.g. Continuing Medical Education (CME) credits.

      (leaf, group, and competency ACTIVE and HISTORY) are assigned to a dedicated Django database
      alias through a database router, mirroring edx-platform's courseware-history router
      (``StudentModuleHistoryExtended``). The alias defaults to the main database, so a stock
      deployment runs everything on one database and gains no new mandatory infrastructure, keeping
      faith with :ref:`openedx-learning-adr-0004`'s no-new-mandatory-infrastructure value; a large
      deployment points the alias at a separate physical database purely through settings, with no
      schema change and no data migration. Building the router now, rather than deferring it, is what
      makes that later split a configuration change instead of a migration on a billion-row table.
      Foreign keys that would cross the alias boundary (a learner-status row referencing a
      criteria-definition row, and the reference to the user) are declared without database-level
      constraints, as edx-platform does for ``PersistentSubsectionGrade.user_id``, so the tables can
      live in different databases; the delete protection of :ref:`openedx-learning-adr-0002` is

>       therefore enforced in application code, not by a database constraint.

Reviewer 3: As mentioned in our discussion on https://github.com/openedx/openedx-core/pull/661 , not having foreign keys is definitely a downside of this approach, and making the deletion protection more complicated and fragile.

> as edx-platform does for ``PersistentSubsectionGrade.user_id``, so the tables can live in different databases

Is there an ADR or documentation around `PersistentSubsectionGrade.user_id` ? I can see that it indeed [is not a foreign key](https://github.com/openedx/openedx-platform/blob/daa2d28e1759d6dd810b4e3cf7ebc802a2310962/lms/djangoapps/grades/models.py#L331), but I can't find any explanation of that in the code nor ADRs, so I'm wondering if it is actually meant to be deployed on a separate database or not. As far as I can tell, the only table that has ever been designed to run on a separate database is the `StudentModuleHistoryExtended` table ([it's the only one with a router](https://github.com/openedx/openedx-platform/blob/daa2d28e1759d6dd810b4e3cf7ebc802a2310962/openedx/envs/common.py#L440-L442)).

Author: @bradenmacdonald @mgwozdz-unicon
I wonder what you think:
- I was trying to model the `StudentModuleHistoryExtended` closely since it already solves the table size issue.
- `StudentCompetencyCriteriaStatus` (the Leaf tables) are the big one. Both ACTIVE and HISTORY tables should be of similar size, which my naive calculations put at worst-case billions of rows for our largest platforms. But could we get away with using the db alias and router for only the history table, if at all? If this is not necessary for the ACTIVE table, we could put a foreign key only on the ACTIVE table, which would be already helpful. I just wonder if there is any similar-sized table on edx-platform that is an example for not using a db router so we know we do not need that optimization?
- I think I was wrong for including `group` and `mastery` here. Those tables (`StudentCompetencyCriteriaGroupStatus`, `CompetencyMasteryStatuses`), as well as their HISTORY tables, are significantly smaller, so I would just use normal foreign keys on them.
- I would love to NOT prematurely optimize here where not necessary. I'm in favor of the most robust and simple solution that is scalable enough. If we can simplify anything here and get rid of anything not needed - all the better.

Author: - `PersistentSubsectionGrade` -> it sounds like I just made a mistake there, I think I should remove that reference.
- Expect me to go through this ADR line by line tomorrow and straightening out misleading references like this and shorten it significantly.

Author: It looks like `StudentModule` is one of the large tables, on a similar magnitude as what we are doing, and is not subject to a db alias and router. I will investigate more to confirm this. Assuming this is correct, I recommend to change this:
- only the `StudentCompetencyCriteriaStatusHistory` - the HISTORY Leaf table - gets the router and db alias, and no foreign keys. All other tables will be normal tables with foreign keys, including the ACTIVE Criteria Status (Leaf table). So we should be mostly free to add foreign keys, with this one exception.
